server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java

```diff
@@ -0,0 +1,175 @@
+                ThreadPool.Names.GENERIC);
+    }
+
+
+    @Override
+    protected void doStop() {
+        scheduledFuture.cancel();
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+
+    }
+
+    public void setEnabled(boolean enabled) { this.enabled = enabled; }
+
+
+    public void setHealthCheckTimeoutInterval(TimeValue healthCheckTimeoutInterval) {
+        this.healthCheckTimeoutInterval = healthCheckTimeoutInterval;
+    }
+
+    @Override
+    public Status getHealth() {
+        if (enabled == false) {
+            return Status.UNKNOWN;
+        }
+        else if ((currentTimeMillisSupplier.getAsLong() - lastRunTimeMillis.get()) >
+            (refreshInterval.millis() + healthCheckTimeoutInterval.millis())) {
+            return Status.UNHEALTHY;
+        }
+        return pathHealthStats.entrySet().stream().anyMatch(map -> map.getValue() == Status.UNHEALTHY) ? Status.UNHEALTHY
+            : Status.HEALTHY;
+    }
+
+     class FsHealthMonitor implements Runnable {
+
+        static final String TEMP_FILE_NAME = ".es_temp_file";
+        private byte[] byteToWrite;
+
+        FsHealthMonitor(){
+            this.byteToWrite = new byte[20];
+        }
+        @Override
+        public void run() {
+            try {
+                if (enabled) {
+                    monitorFSHealth();
+                    logger.debug("Monitor for disk health ran successfully {}", pathHealthStats);
+                }
+            } catch (Exception e) {
+                logger.error("Exception monitoring disk health", e);
+            }
+        }
+
+        private void monitorFSHealth() {
+            if (checkInProgress.compareAndSet(false, true) == false) {
+                logger.warn("Skipping Monitor for disk health as a check is already in progress");
+                return;
+            }
+            for (Path path : nodeEnv.nodeDataPaths()) {
+                try {
+                    if (Files.exists(path)) {
+                        Path tempDataPath = path.resolve(TEMP_FILE_NAME);
+                        Files.deleteIfExists(tempDataPath);
+                        try (OutputStream os = Files.newOutputStream(tempDataPath, StandardOpenOption.CREATE_NEW)) {
+                            new Random().nextBytes(byteToWrite);
+                            os.write(byteToWrite);
+                            IOUtils.fsync(tempDataPath, false);
+                            pathHealthStats.put(path, Status.HEALTHY);
+                        }
+                        Files.delete(tempDataPath);
+                    }
+                } catch (Exception ex) {
+                    logger.error("Failed to perform writes on path {} due to {}", path, ex);
+                    pathHealthStats.put(path,  Status.UNHEALTHY);
+                }
+            }
+            lastRunTimeMillis.getAndUpdate(l -> Math.max(l, currentTimeMillisSupplier.getAsLong()));
```

Comment:
If this individual check saw no exceptions but took longer than the timeout interval then I don't think we should record it as a successful run as is done here, since that will result in a fatally-slow node still occasionally reporting itself as healthy, joining the cluster, and then failing again.

Reply:
Good point. My bad I missed this